### PR TITLE
feat: add BitVec.toNat_[abs|sdiv|smod]

### PR DESCRIPTION
**Pull request by tobiasgrosser**

### CODE REVIEW

File: src/Init/Data/BitVec/Lemmas.lean [hunk 1]
```diff
@@ -1335,6 +1335,15 @@ theorem sdiv_eq (x y : BitVec w) : x.sdiv y =
   rw [BitVec.sdiv]
   rcases x.msb <;> rcases y.msb <;> simp
 
+theorem toNat_sdiv (x y : BitVec w) : (x.sdiv y).toNat =
```
Comment (luisacicolini) [inline src/Init/Data/BitVec/Lemmas.lean:1338]:
```suggestion
theorem toNat_sdiv {x y : BitVec w} : (x.sdiv y).toNat =
```

File: src/Init/Data/BitVec/Lemmas.lean [hunk 2]
```diff
@@ -1358,6 +1367,23 @@ theorem smod_eq (x y : BitVec w) : x.smod y =
   rw [BitVec.smod]
   rcases x.msb <;> rcases y.msb <;> simp
 
+theorem toNat_smod (x y : BitVec w) : (x.smod y).toNat =
```
Comment (luisacicolini) [inline src/Init/Data/BitVec/Lemmas.lean:1370]:
```suggestion
theorem toNat_smod {x y : BitVec w} : (x.smod y).toNat =
```

**Review by luisacicolini — APPROVED** [review]
lgtm

**Review by hargoniX — CHANGES_REQUESTED** [review]
I'm assuming the reason that you didn't put these theorems in the `bv_toNat` simp set is that `omega` cannot handle `match` and `if` yes? Apart from that it looks fine to me.

File: src/Init/Data/BitVec/Lemmas.lean [hunk 3]
```diff
@@ -1335,6 +1335,15 @@ theorem sdiv_eq (x y : BitVec w) : x.sdiv y =
   rw [BitVec.sdiv]
   rcases x.msb <;> rcases y.msb <;> simp
 
+theorem toNat_sdiv {x y : BitVec w} : (x.sdiv y).toNat =
+    match x.msb, y.msb with
+    | false, false => (udiv x y).toNat
+    | false, true  =>  (- (x.udiv (- y))).toNat
+    | true,  false => (- ((- x).udiv y)).toNat
+    | true,  true  => ((- x).udiv (- y)).toNat := by
+  simp [sdiv_eq]
```
[the remark this answers is not in the document]
Comment (tobiasgrosser) [reply src/Init/Data/BitVec/Lemmas.lean]: Done.

File: src/Init/Data/BitVec/Lemmas.lean [hunk 4]
```diff
@@ -1358,6 +1367,23 @@ theorem smod_eq (x y : BitVec w) : x.smod y =
...
+    match x.msb, y.msb with
+    | false, false => (x.umod y).toNat
+    | false, true =>
+      let u := x.umod (- y)
+      (if u = 0#w then u.toNat else (u + y).toNat)
+    | true, false =>
+      let u := (-x).umod y
+      (if u = 0#w then u.toNat else (y - u).toNat)
+    | true, true => (- ((- x).umod (- y))).toNat := by
+  simp [smod_eq]
```
[the remark this answers is not in the document]
Comment (tobiasgrosser) [reply src/Init/Data/BitVec/Lemmas.lean]: Done.

**Comment by tobiasgrosser** [comment]
> I'm assuming the reason that you didn't put these theorems in the `bv_toNat` simp set is that `omega` cannot handle `match` and `if` yes? Apart from that it looks fine to me.

Right. However, on reflection it still makes sense. The pattern will then just be to call bv_toNat, then split on the cases, use omega after.

File: src/Init/Data/BitVec/Lemmas.lean [hunk 5]
```diff
@@ -1961,6 +1989,17 @@ theorem neg_ne_iff_ne_neg {x y : BitVec w} : -x ≠ y ↔ x ≠ -y := by
     subst h'
     simp at h
 
+/-! ### abs -/
+
+@[simp]
```
Comment (hargoniX) [inline src/Init/Data/BitVec/Lemmas.lean:1994]: If you agree that the pattern makes sense, why not `bv_toNat` this one as well?^^

Comment (tobiasgrosser) [reply src/Init/Data/BitVec/Lemmas.lean]: Right. I forgot to push this last change. Its in now.

**Review by hargoniX — APPROVED** [review]

**Pull request merged**